ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -0,0 +1,305 @@
+
+                using cached_program_t =
+                    decltype(program_factory_t::create(operation_attributes, tensor_args, tensor_return_value));
+                auto& cached_program = cached_program_factory.cached_program.template get<cached_program_t>();
+
+                program_factory_t::override_runtime_arguments(
+                    cached_program, operation_attributes, tensor_args, tensor_return_value);
+
+                return cached_program.program;
+            },
+            program_factory);
+        return program;
+    }
+}
+
+template <typename operation_t>
+    requires DeviceOperationConcept<operation_t>
+typename operation_t::tensor_return_value_t run(
+    uint8_t cq_id,
+    const typename operation_t::operation_attributes_t& operation_attributes,
+    const typename operation_t::tensor_args_t& tensor_args) {
+    ZoneScopedN("TT_DNN_DEVICE_OP");
+    uint32_t operation_id = assign_operation_id();
+
+    using tensor_return_value_t = typename operation_t::tensor_return_value_t;
+    static_assert(not std::same_as<tensor_return_value_t, void>, "Operation cannot return type cannot be void");
+
+    auto device = get_first_tensor(tensor_args).get().device();
+    auto& program_cache = device->program_cache;
+
+    auto program_hash = compute_program_hash<operation_t>(operation_attributes, tensor_args);
+    auto cache_hit = program_cache.contains(program_hash);
+
+    if (cache_hit) {
+        operation_t::validate_on_program_cache_hit(operation_attributes, tensor_args);
+    } else {
+        operation_t::validate_on_program_cache_miss(operation_attributes, tensor_args);
+    }
+    auto tensor_return_value = operation_t::create_output_tensors(operation_attributes, tensor_args);
+
+    auto& program = create_or_get_program_from_cache<operation_t>(
+        program_cache, cache_hit, program_hash, operation_attributes, tensor_args, tensor_return_value);
+
+    if (USE_FAST_DISPATCH) {
+        auto& queue = device->command_queue(cq_id);
+        // Program will temporarily own the input buffers. This is required, since with Async command
+        // queues, the input tensor can preemptively be deallocted on device, unless program maintains
+        // explicit ownership. This invocation of the program wicll give up ownership once its enqueued.
+        auto assign_global_buffer_to_program = [&program](auto&& tensor) {
+            AssignGlobalBufferToProgram(tensor.device_buffer(), program);
+        };
+        visit_tensor(assign_global_buffer_to_program, tensor_args);
+        tt::tt_metal::EnqueueProgram(queue, program, false);
+    } else {
+        ::detail::LaunchProgram(device, program);
+    }
+
+    // Visit output tensors with the sole purpose of checking the return type to make sure that it only has Tensors
+    // TODO: come up with a better way of checking the return type
+    visit_tensor([](auto&& tensor) {}, tensor_return_value);
```

Comment:
do we still need this?

Reply:
Yes. we will need this until we write a static check to make sure that the nested types only have `Tensor` in them at the leaf nodes

Reply:
oh i see what you're doing... this will be even faster when we move this out from runtime to compile-time